### ISSUE TITLE
Extracting factor scores and probabilities from multidimensional models

### ISSUE THREAD

**Issue by joakimwallmark** [philchalmers/mirt#236]
I'm trying to compute out of sample accuracy and log-likelihood of fitted IRT models (that is, predictive accuracy and log-likelihood for data not used to fit the model). I'm using the `fscores()` function to get score estimates, and then `probtrace()` with the estimated factor scores to get probabilities for accuracy and log-likelihood computation. 

I am having some issues of 2-factor multidimensional models performing much worse than expected. I am not sure if there is something wrong with my thinking, if I'm using the package in the wrong way or if there is an issue with the package?

Below is a working example with the LSAT7 data for which I'm using the training data to test my own log-likelihood computation. Why is my log-likelihood for the 2 factor model so much larger than the one retrieved from mirt using `logLik()`?

```r
library(mirt)

data <- expand.table(LSAT7)
itemstats(data)

mod1 <- mirt(data, 1)
mod2 <- mirt(data, 2)

mirt_measures <- function(dat, mod, start0 = FALSE, inverse_coded = FALSE) {
  thetas <- fscores(mod, response.pattern = dat)
  probs <- get_mirt_probs(thetas[, grepl("^F", colnames(thetas))], mod)
  log_lik <- acc <- 0
  for (item in seq_along(probs)) {
    if (inverse_coded) { # multidimensional model has revers relationship between prob and score
      dat[, item] <- max(dat[, item]) - dat[, item]
    }
    for (person_id in seq_len(nrow(dat))) {
      if (!is.na(dat[person_id, item])) {
        score_ind <- dat[person_id, item] + 1
        log_lik <- log_lik + log(probs[[item]][person_id, score_ind])
        acc <- acc + (which.max(probs[[item]][person_id, ]) == score_ind)
      }
    }
  }
  acc = acc / (nrow(dat) * length(probs))
  return(c("ll" = log_lik, "acc" = acc))
}

get_mirt_probs <- function(theta, mod) {
  probs <- list()
  for (item in 1:extract.mirt(mod, "nitems")) {
    extr <- extract.item(mod, item)
    probs[[item]] <- probtrace(extr, theta)
  }
  probs
}

logLik(mod1)
logLik(mod2)
mirt_measures(data, mod1)
mirt_measures(data, mod2)
```

**Comment by philchalmers** [comment]
The log-likelihoods are not intended to match. mirt uses MML estimation so the uncertainty in the trait estimates is accounted for via marginalizing over the posterior, while the point estimate approach that your using is more akin to joint ML estimation. Nothing wrong with your code per se, just that these are very different approaches, where MML is notably more consistent.

**Comment by joakimwallmark** [comment]
Thanks a lot for the quick reply. If I understand you correctly, you are saying that the worse performance with my code for 2-dimensional models is likely a consequence of larger uncertainty in the ability estimates from the probtrace() function? Does the logLik function assume the test takers used to fit the model are evenly spread over a joint normal pdf and then compute the likelihood, as part of the MML estimation? 

I wouldn't expect them to match exactly, but I was expecting them to be at least somewhat close if the model has an alright fit. The 1 dimensional model performs somewhat similarly to the MML estimated log-likelihood, while the log-likelihood for the 2D model is much smaller. 

The accuracy is not too far off between the models in my example above, but if I run it with data I did not use to fit the model, I get very bad prediction accuracy and log likelihood compared when using 2D models. As an example, with a large dataset of 60000 people for training from the Swedish SAT test and use 15000 people in my test set to compute log-likelihood and accuracy, I get the following result from a 2pl

|model    |factors| `log likelihood` |accuracy
|-----|-----|-----|-----|
|mirt 2pl |      1|         -732195. |   0.672
|mirt 2pl |      2|         -889393. |   0.578

Using a smaller polytomous dataset (1100 test takers) from national mathematics and data for high school student to fit 1D and 2D GPC models, I get the following

|model|     factors |`log likelihood`| accuracy
|-----|-----|-----|-----|
|mirt gpcm      | 1       |    -5031.    |0.723
|mirt gpcm      | 2       |   -20108.    |0.439

Unfortunately I cannot share my data directly. I am comparing different traditional IRT models with IRT models fit using variational autoencoders as part of my research, and I am just confused by these results. Your package is great overall and it is used to such a wide extent, so I just assumed the fault had to be my own implementation, but maybe these results are to be expected in these situations?
